File: express/src/worker/tasks/email.ts
import { Job, Worker } from "bullmq";
import { Tasks, Queues } from "../../types";
import { emailQueue, config } from "./config";

interface NewsletterData {
    name: string;
    email: string;
}

async function sendNewsletterTask(job: Job<NewsletterData>): Promise<void> {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    const { name, email } = job.data;
    console.log(
        `Newsletter Email sent to ${name} at following address: ${email}`
    );
}

const emailWorker = new Worker(Queues.EMAILS, sendNewsletterTask, config);
emailWorker.on("failed", (job, err) => {
    console.error(`Job ${job?.id} failed with error ${err.message}`);
});

export const sendNewsletter = (name: string, email: string) => {
    emailQueue.add(Tasks.NEWSLETTER, { name, email });
};

File: backend-fastapi/background/tasks/email.py
import time

from background.setup import TASK_NEWSLETTER, NewsletterData, Queues
from background.tasks.broker import dramatiq_task


@dramatiq_task(TASK_NEWSLETTER, Queues.EMAILS)
def send_newsletter_task(data: NewsletterData):
    time.sleep(2)
    print(f"Newsletter sent to {data["name"]} at {data["email"]}")
